### PR TITLE
added hash

### PR DESCRIPTION
**Pull request by eyalk11**
Added hash to Board class

### CODE REVIEW

File: chess/__init__.py [hunk 1]
```diff
@@ -3817,14 +3819,21 @@ def _repr_svg_(self) -> str:
 
     def __eq__(self, board: object) -> bool:
         if isinstance(board, Board):
-            return (
-                self.halfmove_clock == board.halfmove_clock and
-                self.fullmove_number == board.fullmove_number and
-                type(self).uci_variant == type(board).uci_variant and
-                self._transposition_key() == board._transposition_key())
+            return hash(self) == hash(board)
```
Comment (niklasf) [inline chess/__init__.py:3822]: `hash(self) == hash(board)` does *not* imply `self` equals `board` (there can be hash collisions).

**Comment by niklasf** [comment]
From https://docs.python.org/3/reference/datamodel.html#object.__hash__:

> If a class defines mutable objects and implements an [`__eq__()`](https://docs.python.org/3/reference/datamodel.html#object.__eq__) method, it should not implement [`__hash__()`](https://docs.python.org/3/reference/datamodel.html#object.__hash__), since the implementation of [hashable](https://docs.python.org/3/glossary.html#term-hashable) collections requires that a key’s hash value is immutable (if the object’s hash value changes, it will be in the wrong hash bucket).

File: chess/__init__.py (hunk 1 above)
Comment (eyalk11) [reply chess/__init__.py:3822]: I am not sure I understand the claim. Are you claiming the hash is weak? Because hash collisions are negligible.

Comment (niklasf) [reply chess/__init__.py:3822]: Maybe for your use case.

But this is a general purpose library. So we shouldn't assume that we will only ever see 2^32 positions (at which point collisions are likely even for strong 64 bit hashes), or even that inputs aren't specifically chosen by an adversary.

**Pull request closed without merging**